### PR TITLE
Support multiple Proxy deployments in the same transaction

### LINKED ISSUE #2250: L2: Support multiple Proxy deployments in the same transaction
Currently if there are more than one Proxy deployment in the same transaction they will not be decoded propertly.

There are some ways this can be done for proxies deployed using our official `ProxyFactory` :

- Check `ProxyCreation` events on the transaction and match the order with the list of decoded data.
- For `saltNonce` deployments calculate the deployment address using the decoded `saltNonce` and `initializer`.

### PR DESCRIPTION
**Pull request by Uxio0**
- Closes #2250

### CODE REVIEW

**Review by moisses89 — CHANGES_REQUESTED** [review]

File: safe_transaction_service/history/models.py [hunk 1]
```diff
@@ -452,6 +453,18 @@ def update_with_block_and_receipt(
                 ]
             )
 
+    def get_deployed_proxies_from_logs(self):
+        """
+        :return: list of `SafeProxyFactory` proxies that emitted the `ProxyCreation` event on this transaction
+        """
+        return [
+            fast_to_checksum_address(HexBytes(log["topics"][1])[12:])
```
[the remark this answers is not in the document]
Comment (Uxio0) [reply safe_transaction_service/history/models.py:461]: I don't think there's an easier way to do that, but I will add the comment

File: safe_transaction_service/history/models.py [hunk 2]
```diff
@@ -452,6 +453,18 @@ def update_with_block_and_receipt(
                 ]
             )
 
+    def get_deployed_proxies_from_logs(self):
```
[the remark this answers is not in the document]
Comment (Uxio0) [reply safe_transaction_service/history/models.py:456]: Nice one

File: safe_transaction_service/history/models.py [hunk 3]
```diff
@@ -452,6 +453,18 @@ def update_with_block_and_receipt(
...
+        :return: list of `SafeProxyFactory` proxies that emitted the `ProxyCreation` event on this transaction
+        """
+        return [
+            fast_to_checksum_address(HexBytes(log["topics"][1])[12:])
+            for log in self.logs
+            if log["topics"]
+            and len(log["topics"]) == 2
+            and HexBytes(log["topics"][0]) == SAFE_PROXY_FACTORY_CREATION_EVENT_TOPIC
+        ]
+
```
[the remark this answers is not in the document]
Comment (Uxio0) [reply safe_transaction_service/history/models.py:467]: No. the topic zero is the "signature" of the event. In this case I think there's no need for comments as this is EVM knowledge, and it's already the basis of the indexer

File: safe_transaction_service/history/services/safe_service.py [hunk 4]
```diff
@@ -217,6 +224,36 @@ def get_safe_info_from_db(self, safe_address: ChecksumAddress) -> SafeInfo:
...
+
+        if not results:
+            return None
+
+        if len(results) == 1:
+            return results[0]
+
+        # If there are more than one deployment, we need to know which one is the one we need
+        deployed_safes = ethereum_tx.get_deployed_proxies_from_logs()
+        if len(deployed_safes) == len(results):
```
[the remark this answers is not in the document]
Comment (Uxio0) [reply safe_transaction_service/history/services/safe_service.py:251]: Do you mean when will it NOT happen?

File: safe_transaction_service/history/models.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Uxio0) [reply safe_transaction_service/history/models.py:467]: https://docs.soliditylang.org/en/v0.8.17/contracts.html#events

File: safe_transaction_service/history/models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Uxio0) [reply safe_transaction_service/history/models.py:463]: Done

File: safe_transaction_service/history/models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Uxio0) [reply safe_transaction_service/history/models.py]: Done

File: safe_transaction_service/history/services/safe_service.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Uxio0) [reply safe_transaction_service/history/services/safe_service.py:251]: Added a log for that

**Review by moisses89 — APPROVED** [review]

**Pull request merged**